rust_src/src/buffers.rs

```diff
@@ -1014,6 +1017,51 @@ pub fn erase_buffer() {
     }
 }
 
+/// Return a string that is the name of no existing buffer based on NAME.
+/// If there is no live buffer named NAME, then return NAME.
+/// Otherwise modify name by appending `<NUMBER>', incrementing NUMBER
+/// (starting at 2) until an unused name is found, and then return that name.
+/// Optional second argument IGNORE specifies a name that is okay to use (if
+/// it is in the sequence to be tried) even if a buffer with that name exists.
+///
+/// If NAME begins with a space (i.e., a buffer that is not normally
+/// visible to users), then if buffer NAME already exists a random number
+/// is first appended to NAME, to speed up finding a non-existent buffer.
+#[lisp_fn(min = "1")]
+pub fn generate_new_buffer_name(name: LispStringRef, ignore: LispObject) -> LispStringRef {
+    if get_buffer(LispBufferOrName::Name(name.into())).is_none()
+        || (ignore != Qnil && string_equal(name.into(), ignore))
+    {
+        return name;
+    }
+
+    let basename = if name.byte_at(0) == b' ' {
+        let range = Range::new(0, 1000000);
+        let mut rng = rand::thread_rng();
+        // Since random only contains ascii chars, we can use
+        // mock_unibyte_string
+        let suffix = mock_unibyte_string!(format!("-{}", range.ind_sample(&mut rng)));
```

Comment:
Don't use `mock_unibyte_string` here. It's only intended to be used for testing, since the "string" that it creates doesn't participate in the Lisp garbage collector.

Reply:
It is intended to be that way. C code too allocates the Lisp_String on the stack if possible. mock_unibyte_string too does the same.

Reply:
If that is the case for sure, we should come up with a generic "local allocator" which the mock system should call.

Reply:
Currently the way to initialize the object is specific to the object's type. I'm not sure if we can generalize the allocator.